template/flake.nix

```diff
@@ -2,38 +2,80 @@
   description = "Your jupyterWith project";
 
   inputs.nixpkgs.follows = "jupyterWith/nixpkgs";
+  inputs.flake-compat.url = "github:edolstra/flake-compat";
+  inputs.flake-compat.flake = false;
   inputs.flake-utils.url = "github:numtide/flake-utils";
   inputs.jupyterWith.url = "github:tweag/jupyterWith";
 
-  # inputs.jupyterWith.inputs.nixpkgs.follows = "nixpkgs";
-
   outputs = {
     self,
     nixpkgs,
+    flake-compat,
     flake-utils,
     jupyterWith,
   }:
+  # TODO - Update Linux first and then MacOS when it is working.
     flake-utils.lib.eachSystem ["x86_64-linux"] (system: let
+      inherit (jupyterWith.lib.${system}) mkKernel mkJupyterlabInstance;
+
       pkgs = import nixpkgs {
         inherit system;
-        overlays = builtins.attrValues jupyterWith.overlays;
+        overlays = [
+          self.overlays.default
+        ];
       };
 
-      kernels = let
-        inherit (builtins) map readDir attrNames;
+      jupyterEnvironment = let
+        inherit (builtins) listToAttrs map readDir attrNames;
+        inherit (pkgs.lib) removeSuffix;
         inherit (pkgs.lib.attrsets) filterAttrs;
         inherit (pkgs.lib.strings) hasPrefix hasSuffix;
-      in
-        map (name: import ./kernels/${name} {inherit pkgs;})
-        (attrNames
-          (filterAttrs
-            (n: v: v == "regular" && hasSuffix ".nix" n && !hasPrefix "_" n)
-            (readDir ./kernels)));
 
-      jupyterEnvironment =
-        pkgs.jupyterWith.jupyterlabWith {inherit kernels;};
+        /*
+        Takes a file name, `name` and a file type, `value`, and returns a
+        boolean if the file is meant to be an available kernel. Kernels whose
+        file names are prefixed with an underscore are meant to be hidden.
+        Useful for filtering the output of `readDir`.
+        */
+        filterAvailableKernels = name: value:
+          (value == "regular")
+          && hasSuffix ".nix" name
+          && !hasPrefix "_" name;
+
+        /*
+        Gets the available kernels as a set from the kernels directory. Name is
+        the kernel name and value is the file type.
+        */
+        getAvailableKernels =
+          filterAttrs
+          filterAvailableKernels
+          (readDir ./kernels);
+
+        /*
+        Takes set of kernels, `kernels`, and a kernel name, `name`,
+        and imports it from the kernels directory.
+        Returns the imported kernels as the value of an attribute set.
+        */
+        importKernel = kernels: name: {
+          name = removeSuffix ".nix" name;
+          value = import ./kernels/${name} {inherit pkgs name mkKernel kernels;};
+        };
+      in
+        mkJupyterlabInstance {
+          kernels = kernels:
+            listToAttrs (
```

Comment:
move all the extra code to `jupyterWith` and than import it like `mkKernel` function. example: `readKernelsFromPath pkgs ./kernels` could be an example